panic-rtt-target/src/lib.rs

```diff
@@ -50,29 +50,29 @@ use core::{
 #[allow(unused)]
 use rtt_target::{ChannelMode, UpChannel};
 
-#[cfg(feature = "cortex-m")]
 #[inline(never)]
 #[panic_handler]
 fn panic(info: &PanicInfo) -> ! {
-    use cortex_m::interrupt;
+    critical_section::with(|_| {
+        if let Some(mut channel) = unsafe { UpChannel::conjure(0) } {
+            channel.set_mode(ChannelMode::BlockIfFull);
 
-    interrupt::disable();
+            writeln!(channel, "{}", info).ok();
+        } else {
+            // failed to get channel, but not much else we can do but spin
+            loop {
+                compiler_fence(SeqCst);
+            }
+        }
 
-    if let Some(mut channel) = unsafe { UpChannel::conjure(0) } {
-        channel.set_mode(ChannelMode::BlockIfFull);
-
-        writeln!(channel, "{}", info).ok();
-    }
+        // we should never leave critical section
+        loop {
+            compiler_fence(SeqCst);
+        }
+    });
 
+    // will in fact be unreachable
     loop {
```

Comment:
I think this one should be removable as well, if you keep the one inside

Reply:
Indeed. I had a compilation error, forcing me to have the extra loop, but I figured it out now. 
TL;DR, I made a stupid mistake :)

I had a `;` after the `critical_section::with(....);` <- here.
Even if Rust is smart enough to figure out that the closure is a ! (and as such would never terminate), Rust type checker goes by concrete types (not inference). thus by removing the `;` the inferred type `!` gets propagated. In this case a "smarter" compiler would have propagate the `effect` of not returning from the closure, but that was not the case.